MapboxCoreNavigation/RouteController.swift

```diff
@@ -322,26 +322,25 @@ open class RouteController: NSObject {
         let relativeAnglepointBehind = wrap(wrappedPointBehind - wrappedCourse, min: -180, max: 180)
         let relativeAnglepointAhead = wrap(wrappedPointAhead - wrappedCourse, min: -180, max: 180)
         let averageRelativeAngle = (relativeAnglepointBehind + relativeAnglepointAhead) / 2
-        let absoluteDirection = wrap(wrappedCourse + averageRelativeAngle, min: 0 , max: 360)
+        let calculatedCourseForLocationOnStep = wrap(wrappedCourse + averageRelativeAngle, min: 0 , max: 360)
         
-        // If the course is inaccurate or the speed is low and the user is on the route,
-        // snap the users location and course since we know the snapped location and course is more accurate.
-        if location.course <= 0 || location.speed <= RouteControllerMinimumSpeedThresholdForSnappingUserToRoute, snappedCoordinate.distance < RouteControllerUserLocationSnappingDistance {
-            let calculatedWrappedCourse = wrap((wrappedPointBehind + wrappedPointAhead) / 2, min: 0 , max: 360)
-            return CLLocation(coordinate: snappedCoordinate.coordinate, altitude: location.altitude, horizontalAccuracy: location.horizontalAccuracy, verticalAccuracy: location.verticalAccuracy, course: calculatedWrappedCourse, speed: location.speed, timestamp: location.timestamp)
-        }
-
-        guard differenceBetweenAngles(absoluteDirection, location.course) < RouteControllerMaxManipulatedCourseAngle else {
-            return location
-        }
-
-        let course = averageRelativeAngle <= RouteControllerMaximumAllowedDegreeOffsetForTurnCompletion ? absoluteDirection : location.course
+        var userCourse = calculatedCourseForLocationOnStep
+        var userCoordinate = snappedCoordinate.coordinate
+        
+        print("course:", location.course, "speed:", location.speed, "distance from route:", snappedCoordinate.distance, "accuracy:", location.horizontalAccuracy, "difference between expected course:", differenceBetweenAngles(calculatedCourseForLocationOnStep, location.course))
         
-        guard snappedCoordinate.distance < RouteControllerUserLocationSnappingDistance else {
-            return location
+        
+        if location.course >= 0 || location.speed >= RouteControllerMinimumSpeedThresholdForSnappingUserToRoute {
+            if differenceBetweenAngles(calculatedCourseForLocationOnStep, location.course) > RouteControllerMaxManipulatedCourseAngle && location.horizontalAccuracy < 20 {
+                userCourse = location.course
+                
+                if snappedCoordinate.distance > RouteControllerUserLocationSnappingDistance && location.horizontalAccuracy < 20 {
```

Comment:
You can only have the actual raw location if the user course is also different than what we think it is.